core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java

```diff
@@ -280,16 +285,19 @@ protected RelNode decorrelate(RelNode root) {
     final Frame frame = getInvoke(root, null);
     if (frame != null) {
       // has been rewritten; apply rules post-decorrelation
-      final HepProgram program2 = HepProgram.builder()
+      final HepProgramBuilder builder = HepProgram.builder()
           .addRuleInstance(
               CoreRules.FILTER_INTO_JOIN.config
                   .withRelBuilderFactory(f)
                   .toRule())
           .addRuleInstance(
               CoreRules.JOIN_CONDITION_PUSH.config
                   .withRelBuilderFactory(f)
-                  .toRule())
-          .build();
+                  .toRule());
+      if (getPostDecorrelateRules().size() > 0) {
```

Comment:
It may be more efficient to replace this with `!getPostDecorrelateRules().isEmpty()`

Reply:
Which kind of list has a more efficient impl ?

Reply:
For example, ArrayList has O(1) time complexity for `size()` operation. However, the time complexity for org.apache.calcite.runtime.ConsList is O(n).